### PR TITLE
Mark existing views as distributed when upgrade to 11.0+

### PR DESCRIPTION
**Pull request by onderkalaci**
We have a mechanism which ensures that newly distributed
objects are recorded during `alter extension citus update`.

However, the logic was lacking "view"s. With this commit, we make
sure that existing views are also marked as distributed during
upgrade.

On top of #5914 (and #5889) for ease of testing.

### CODE REVIEW

File: src/backend/distributed/commands/extension.c [hunk 1]
```diff
@@ -515,19 +515,70 @@ MarkExistingObjectDependenciesDistributedIfSupported()
 
 		if (ShouldSyncTableMetadata(citusTableId))
 		{
-			/* we need to pass pointer allocated in the heap */
-			ObjectAddress *addressPointer = palloc0(sizeof(ObjectAddress));
-			*addressPointer = tableAddress;
+			/*
```
Comment (onderkalaci) [inline src/backend/distributed/commands/extension.c:518]: Probably a separate thing to look. Considering https://github.com/citusdata/citus/pull/4344, maybe we should use `ShouldSyncTableMetadataViaCatalog` instead of `ShouldSyncTableMetadata` ?

Worth testing an upgrade with 100k distributed tables.

Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c:518]: Not super critical impact, but still I'd be in favor of merging this https://github.com/citusdata/citus/pull/5925

Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c:518]: already merged

File: src/backend/distributed/commands/extension.c [hunk 2]
```diff
@@ -515,19 +515,70 @@ MarkExistingObjectDependenciesDistributedIfSupported()
...
+			 * as distributed while it is processed for the last citus
+			 * table.
+			 */
+			MarkObjectDistributedLocally(&tableAddress);
+
+			/*
+			 * As of Citus 11, views on Citus tables that do not have any unsupported
+			 * dependency should also be distributed.
+			 */
+			List *viewList = GetDependingViews(citusTableId);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c:555]: we skip && added test

File: src/backend/distributed/commands/extension.c [hunk 3]
```diff
@@ -515,19 +515,70 @@ MarkExistingObjectDependenciesDistributedIfSupported()
...
+				 * well. For example, a type could only be depending to a
+				 * view.
+				 */
+
+				/* we need to pass pointer allocated in the heap */
+				ObjectAddress *addressPointer = palloc0(sizeof(ObjectAddress));
+				*addressPointer = viewAddress;
+
+				/* */
+				resultingObjectAddresses =
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c:593]: we skip views/tables owned by extensions.

File: src/backend/distributed/commands/extension.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (velioglu) [reply src/backend/distributed/commands/extension.c:556]: Since we don't add support for mat views, it is ok to skip them.

File: src/backend/distributed/commands/extension.c [hunk 4]
```diff
@@ -510,25 +510,98 @@ MarkExistingObjectDependenciesDistributedIfSupported()
 	Oid citusTableId = InvalidOid;
 	foreach_oid(citusTableId, citusTableIdList)
 	{
+		if (IsTableOwnedByExtension(citusTableId))
```
Comment (velioglu) [inline src/backend/distributed/commands/extension.c:513]: I think we should have done that even before view, but we don't need to backport it to the 11 Beta as we'll provide complete support with 11, right?

File: src/backend/distributed/commands/extension.c [hunk 5]
```diff
@@ -510,25 +510,98 @@ MarkExistingObjectDependenciesDistributedIfSupported()
...
+			 *
+			 * If a view depends on multiple tables, that view will be marked
+			 * as distributed while it is processed for the last citus
+			 * table.
+			 */
+			MarkObjectDistributedLocally(&tableAddress);
 
-		resultingObjectAddresses = list_concat(resultingObjectAddresses,
-											   distributableDependencyObjectAddresses);
+			/*
```
Comment (velioglu) [inline src/backend/distributed/commands/extension.c:552]: We also support distributing views like `CREATE VIEW basic_view AS SELECT 1;`. In addition to obtaining views depending on any Citus table, should we go over all the views and distribute ones don't depend on any table with an additional loop? Or should we completely skip them to keep code simple?

File: src/backend/distributed/metadata/metadata_cache.c [hunk 6]
```diff
@@ -2506,6 +2507,17 @@ DistPlacementPlacementidIndexId(void)
 }
 
 
+/* return oid of pg_dist_colocation_pkey */
+Oid
+DistColocationIndexId(void)
```
Comment (velioglu) [inline src/backend/distributed/metadata/metadata_cache.c:2512]: Could you please make colocation index related code a separate commit? It is not so easy to understand that we've added them for the sake of test consistency now.

File: src/test/regress/sql/upgrade_post_11_before.sql [hunk 7]
```diff
@@ -129,6 +142,68 @@ END;
 $$;');
 
 CREATE TYPE post_11_upgrade.my_type AS (a int);
+CREATE VIEW post_11_upgrade.view_for_upgrade_test AS SELECT * FROM sensors;
```
Comment (velioglu) [inline src/test/regress/sql/upgrade_post_11_before.sql:145]: Once you add support for basic views like `CREATE VIEW basic_view AS SELECT 1` could you also please add a related test as well?

File: src/backend/distributed/commands/extension.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c:513]: Hmm, actually backporting this pieces could make sense, though not urgent.

So, I can create a separate commit and cherry-pick to 10.0+.

Does that sound good?

File: src/backend/distributed/metadata/metadata_cache.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_cache.c:2512]: makes sense. I'll do that once the PR is approved

File: src/backend/distributed/commands/extension.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/extension.c]: Yes, we should probably support. I pushed a refactor now, and working on adding some tests / discovering whether any further edge cases exists with this approach.

Do you see any?

File: src/test/regress/sql/upgrade_post_11_before.sql (hunk 7 above)
Comment (onderkalaci) [reply src/test/regress/sql/upgrade_post_11_before.sql:145]: added

**Review by velioglu — APPROVED** [review]
